### PR TITLE
RD-5720: Migrated form components to typescript - Part 1

### PR DESCRIPTION
**Pull request by Vorbert-Kruk**
## Description

Changes consists of:
- fully migrating inputs (related to time from `src/components/form` directory) to typescript
- migrating away from implicit lodash imports, in order of using explicit ones
- removing redundant `@ts-expect-error` comments 
- adjusting html snapshots used by tests
- fixing displaying stories (DataInput component was not being displayed - for more details [click here](https://github.com/cloudify-cosmo/cloudify-ui-components/pull/168/files#r992485606))
- marked`@homemx/react-live` as `devDependencies`
- resolve typing inconsistency for `PopupConfirm` - for more details [click here](https://github.com/cloudify-cosmo/cloudify-ui-components/pull/168/files#r995568015)

## Screenshots / Videos

N/A

## Checklist

- [x] My code follows the [UI Code Style](https://cloudifysource.atlassian.net/l/c/x8fq902p).
- [x] I followed the [UI Testing Guidelines](https://cloudifysource.atlassian.net/l/cp/udCVfvrx) and verified that all tests/checks have passed.
- [x] I verified if that change requires any update in the [documentation](https://cloudifysource.atlassian.net/l/c/4XKtPocy).
- [x] I added proper labels to this PR.

## Tests

 

N/A

## Documentation

N/A

### CODE REVIEW

File: src/components/form/DateRangeInput/DateRangeInput.tsx [hunk 1]
```diff
@@ -398,106 +488,4 @@ export default class DateRangeInput extends React.PureComponent {
...
-// @ts-expect-error TS(2339) FIXME: Property 'initialState' does not exist on type 'ty... Remove this comment to see the full error message
-DateRangeInput.initialState = props => ({
-    ...props.defaultValue, // { range, start, end }
-    ...props.value, // { range, start, end }
-    ...DateRangeInput.getStartDateState(props.value.start), // { startError, startDate }
-    ...DateRangeInput.getEndDateState(props.value.end), // { endError, endDate }
-    isOpen: props.defaultOpen,
-    dirty: false
-});
+export default DateRangeInput;
```
Comment (Vorbert-Kruk) [inline src/components/form/DateRangeInput/DateRangeInput.tsx:491]: ### Note
Component has been exported like this due to the static properties issue, mentioned and explained [here](https://github.com/storybookjs/storybook/issues/11419#issuecomment-658969643)

File: test/DateRangeInput.test.tsx [hunk 2]
```diff
@@ -19,7 +19,6 @@ describe('<DateRangeInput />', () => {
     it('renders', () => {
         const wrapper = mount(
             <DateRangeInput
-                // @ts-expect-error TS(2322) FIXME: Type '{ name: string; value: { range: string; star... Remove this comment to see the full error message
```
Comment (Vorbert-Kruk) [inline test/DateRangeInput.test.tsx:22]: ### Note
I did not wanted to migrate test file as it's not in the scope of the ticket
But as I needed to adjust some parts of it, in order of progressing with the migration, I removed some of the errors 🎉

File: src/components/popups/PopupConfirm/PopupConfirm.tsx [hunk 3]
```diff
@@ -19,7 +19,7 @@ export interface PopupConfirmProps extends PopupProps {
     /**
      * function called to determine if Cancel button should be displayed
      */
-    onCanConfirm?: () => string | undefined;
+    onCanConfirm?: () => string | void;
```
Comment (Vorbert-Kruk) [inline src/components/popups/PopupConfirm/PopupConfirm.tsx:22]: ### Note
Typing provided within #169 did not fully resolve an issue (which is described in #169)
This modification should fully resolve the issue 💪

File: test/__snapshots__/HtmlSnapshots.test.ts.snap [hunk 4]
```diff
@@ -10472,7 +10472,6 @@ exports[`Storyshots Form/DateRange Basic 1`] = `
     onFocus={[Function]}
     onMouseEnter={[Function]}
     onMouseLeave={[Function]}
-    placeholder=""
```
Comment (Vorbert-Kruk) [inline test/__snapshots__/HtmlSnapshots.test.ts.snap:10475]: ### Note
It's being removed due to replacing default prop from `""` to `undefined`
It doesn't make any difference from a functionality perspective 💪

File: src/components/form/DateRangeInput/DateRangeInput.tsx [hunk 5]
```diff
@@ -1,14 +1,109 @@
...
+import type { DatePickerProps } from '../DatePicker/DatePicker';
+
+interface DateRangeValue {
+    range: string;
+    start: string;
+    end: string;
+}
+
+type RangeValue = Omit<DateRangeValue, 'range'>;
+type Range = Record<string, RangeValue>;
```
Comment (qooban) [inline src/components/form/DateRangeInput/DateRangeInput.tsx:24]: I'm wondering if we can improve the naming here 🤔 

What do you think about that:
```suggestion
interface Range {
    start: string;
    end: string;
}

interface DateRange extends Range {
    range: string;
}

type Ranges = Record<string, Range>;
```
?

File: src/components/form/DateInput/DateInput.tsx [hunk 6]
```diff
@@ -1,86 +1,163 @@
...
 
+type OnChangeEvent =
+    | Parameters<Required<DatePickerProps>['onChange']>[0]
+    | Parameters<Required<InputProps>['onChange']>[0]
+    | Parameters<Required<ButtonProps>['onClick']>[0];
+
+interface onChangeData {
+    name: string;
+    value: string;
+}
```
Comment (qooban) [inline src/components/form/DateInput/DateInput.tsx:20]: Possibly should be exported to be available for Stage/Composer? 🤔

File: test/DateRangeInput.test.tsx (hunk 2 above)
Comment (Vorbert-Kruk) [reply test/DateRangeInput.test.tsx:22]: Oh, that one I've missed
Migrating it right away! 😁

Comment (Vorbert-Kruk) [reply test/DateRangeInput.test.tsx:22]: Done 🎉

File: src/components/form/TimeInput/TimeInput.tsx [hunk 7]
```diff
@@ -1,40 +1,57 @@
...
         return { key: value, text: value, value };
     });
 };
 const hoursOptions = generateOptions(0, 24, 1, 2);
 const minutesOptions = generateOptions(0, 60, 1, 2);
 
+interface onChangeData {
+    name: string;
+    value: string;
+}
```
[the remark this answers is not in the document]
Comment (Vorbert-Kruk) [reply src/components/form/TimeInput/TimeInput.tsx:19]: Ad 1.
Yeah, my bad 😁 - I'll adjust it right away 💪 

Ad 2.
Good Idea, surely I'll do so 😉 
And if we'd want to use them in the future, I'll modify their name to be consisting of `${ComponentName}OnChangeDate` 🎉 

Ad 3.
I was thinking about it, but as 2 other components were implementing this type in a different way I thought that it didn't bring us much of a value
So if that'd be an option, I'll leave it as it is

[the remark this answers is not in the document]
Comment (Vorbert-Kruk) [reply src/components/form/TimeInput/TimeInput.tsx]: Changes were provided 🚀

File: src/components/form/DateInput/DateInput.tsx [hunk 8]
```diff
@@ -1,86 +1,163 @@
...
+    dateValue?: null | Moment;
+    dateError: boolean;
+    dirty: boolean;
+}
+
+const initialState: DateInputState = {
+    dateError: false,
+    dateValue: null,
+    dirty: false
+};
```
[the remark this answers is not in the document]
Comment (Vorbert-Kruk) [reply src/components/form/DateInput/DateInput.tsx]: No, we don't need it 😁 
Nice catch!

I've removed it 💪

File: src/components/form/DateInput/DateInput.tsx [hunk 9]
```diff
@@ -1,86 +1,163 @@
...
+    };
+
+    static initialState = {
+        dateError: false,
+        dateValue: null,
+        dirty: false
+    };
+
+    constructor(props: DateInputProps, context: unknown) {
         super(props, context);
```
[the remark this answers is not in the document]
Comment (Vorbert-Kruk) [reply src/components/form/DateInput/DateInput.tsx]: Hmm, let's see 😁 
I've removed it from both `DateInput` and `DateRangeInput` 💪

File: src/components/form/DateRangeInput/DateRangeInput.tsx [hunk 10]
```diff
@@ -1,14 +1,109 @@
...
+
+    /**
+     * text to be displayed in the input field when there's no range set
+     */
+    placeholder?: string;
+
+    /**
+     * ranges object containing dict of named custom ranges with start/end string dates "YYYY-MM-DD HH:mm"
+     */
+    ranges?: Range;
```
[the remark this answers is not in the document]
Comment (Vorbert-Kruk) [reply src/components/form/DateRangeInput/DateRangeInput.tsx]: Good point 
I've adjusted it 🎉

File: src/components/form/DateInput/DateInput.tsx (hunk 6 above)
Comment (Vorbert-Kruk) [reply src/components/form/DateInput/DateInput.tsx]: Did as a part of a different comment
We can see the modification as a part of [this commit](https://github.com/cloudify-cosmo/cloudify-ui-components/pull/168/commits/21935d83bf67122998c7a14b4597751ce03aa3bf) 🚀

File: src/components/form/DateRangeInput/DateRangeInput.tsx (hunk 5 above)
Comment (Vorbert-Kruk) [reply src/components/form/DateRangeInput/DateRangeInput.tsx]: Sure, seems like a good suggestion :)

Comment (Vorbert-Kruk) [reply src/components/form/DateRangeInput/DateRangeInput.tsx]: Done as a part of [this commit](https://github.com/cloudify-cosmo/cloudify-ui-components/pull/168/commits/ce681665096cbf4e5c18148217a3f81a4457e4d6)

File: src/components/form/DateRangeInput/DateRangeInput.tsx [hunk 11]
```diff
@@ -1,14 +1,109 @@
...
+
+interface EndDate {
+    endError: boolean;
+    endDate?: Moment;
+}
+
+interface onChangeData {
+    name: string;
+    value: DateRangeValue;
+}
```
[the remark this answers is not in the document]
Comment (Vorbert-Kruk) [reply src/components/form/DateRangeInput/DateRangeInput.tsx]: Sure, nicely thought
It has been done as a part of [this commit](https://github.com/cloudify-cosmo/cloudify-ui-components/pull/168/commits/c85d78b0285162aad3e088aedfd39e5cc93f5bf1)

**Review by qooban — APPROVED** [review]
Great piece of work 👍🏻

**Pull request merged**